controllers/clustergroupupgrade_controller.go

```diff
@@ -1662,62 +1662,64 @@ func (r *ClusterGroupUpgradeReconciler) checkDuplicateChildResources(ctx context
 				r.Log.Info("[checkDuplicateChildResources] clean up stale child resource", "name", u.GetName(), "kind", u.GetKind())
 				err := r.Client.Delete(ctx, u)
 				if err != nil {
-					return childResourceNameList, err
+					return childResourceNames, err
 				}
-				return childResourceNameList, nil
+				return childResourceNames, nil
 			}
 		} else {
 			safeNameMap[desiredName] = u.GetName()
 		}
 	}
-	childResourceNameList = append(childResourceNameList, u.GetName())
-	return childResourceNameList, nil
+	childResourceNames = append(childResourceNames, u.GetName())
+	return childResourceNames, nil
 }
 
-func (r *ClusterGroupUpgradeReconciler) updateStatus(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+func (r *ClusterGroupUpgradeReconciler) updateChildResourceNamesInStatus(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
```

Comment:
I am thinking why we cannot update the CGU status after each resource is created instead of querying afterwards

ie. update the copiedPolicies list after the r.Create() is succeeded for the new policy

clusterGroupUpgrade.Status.CopiedPolicies = append(clusterGroupUpgrade.Status.CopiedPolicies, policyname)

Reply:
I don't see why not either if we didn't do the random name suffix. With the new name mapping, we actually need to do a query in case multiple copies get created. It can happen when reconcile call gets a stale version of CGU.

Reply:
It's already done that way. The case here is that the CGU we get at the start can be a stale copy, when a reconcile event happens right after the previous call completes. If the previous call is the one that did the resource creation, we end up creating duplicate copies as the map in the stale copy is still empty. The reason for the stale copy is due to the eventual consistency model of the etcd replication. To minimize the chance of this, I added a bit delay at the start. But it could still happen if the replication op happens to be extra slow.

Reply:
Got you. The stale version of CGU could have totally wrong information including empty safeResourceNames map.  Thanks for explaining!

Reply:
```
2022-06-03T10:52:54.223-0400	INFO	controllers.ClusterGroupUpgrade	Start reconciling CGU	{"name": "cnfdf18-new", "version": "189219603"}
2022-06-03T10:52:54.223-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.223-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.298-0400	INFO	controllers.ClusterGroupUpgrade	Start reconciling CGU	{"name": "cnfdf18-new", "version": "189219604"}
2022-06-03T10:52:54.298-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.298-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.359-0400	INFO	controllers.ClusterGroupUpgrade	Finish reconciling CGU	{"name": "cnfdf18-new"}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	Start reconciling CGU	{"name": "cnfdf18-new", "version": "189219611"}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.369-0400	INFO	controllers.ClusterGroupUpgrade	[doManagedPoliciesExist]	{"policyMap": {"common-cnfdf18-subscriptions-policy":"ztp-common-cnfdf18"}}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersNonCompliantWithPolicy]	{"policy: ": "common-cnfdf18-subscriptions-policy", "clusters: ": ["cnfdf18"]}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.427-0400	INFO	controllers.ClusterGroupUpgrade	Remediation plan	{"remediatePlan": [["cnfdf18"]]}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:54.958-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersNonCompliantWithPolicy]	{"policy: ": "common-cnfdf18-subscriptions-policy", "clusters: ": ["cnfdf18"]}
2022-06-03T10:52:54.958-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV doesn't exist, create it	{"name": "cnfdf18-new-ztp-install-subscription-sriov-network-operator-subscription", "namespace": "cnfdf18"}
2022-06-03T10:52:55.022-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV doesn't exist, create it	{"name": "cnfdf18-new-ztp-install-subscription-ptp-operator-subscription", "namespace": "cnfdf18"}
2022-06-03T10:52:55.081-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV doesn't exist, create it	{"name": "cnfdf18-new-ztp-install-subscription-performance-addon-operator", "namespace": "cnfdf18"}
2022-06-03T10:52:55.141-0400	INFO	controllers.ClusterGroupUpgrade	[blockingCRsNotCompleted]	{"blockingCRs": []}
2022-06-03T10:52:55.141-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.141-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.235-0400	INFO	controllers.ClusterGroupUpgrade	[Reconcile]	{"RequeueAfter:": "5m0s"}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	Finish reconciling CGU	{"name": "cnfdf18-new"}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	Start reconciling CGU	{"name": "cnfdf18-new", "version": "189219611"}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.297-0400	INFO	controllers.ClusterGroupUpgrade	[doManagedPoliciesExist]	{"policyMap": {"common-cnfdf18-subscriptions-policy":"ztp-common-cnfdf18"}}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersNonCompliantWithPolicy]	{"policy: ": "common-cnfdf18-subscriptions-policy", "clusters: ": ["cnfdf18"]}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.354-0400	INFO	controllers.ClusterGroupUpgrade	Remediation plan	{"remediatePlan": [["cnfdf18"]]}
2022-06-03T10:52:55.889-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind is not of Subscription kind	{"policyName": "common-cnfdf18-subscriptions-policy"}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:55.890-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersNonCompliantWithPolicy]	{"policy: ": "common-cnfdf18-subscriptions-policy", "clusters: ": ["cnfdf18"]}
2022-06-03T10:52:55.890-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV already exists, update it	{"name": "cnfdf18-new-ztp-install-subscription-sriov-network-operator-subscription", "namespace": "cnfdf18"}
2022-06-03T10:52:55.952-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV already exists, update it	{"name": "cnfdf18-new-ztp-install-subscription-ptp-operator-subscription", "namespace": "cnfdf18"}
2022-06-03T10:52:56.009-0400	INFO	multiCloudLog	[EnsureManagedClusterView] MCV already exists, update it	{"name": "cnfdf18-new-ztp-install-subscription-performance-addon-operator", "namespace": "cnfdf18"}
2022-06-03T10:52:56.067-0400	INFO	controllers.ClusterGroupUpgrade	[blockingCRsNotCompleted]	{"blockingCRs": []}
2022-06-03T10:52:56.067-0400	INFO	controllers.ClusterGroupUpgrade	[getClusterBySelectors]	{"clustersBySelector": []}
2022-06-03T10:52:56.067-0400	INFO	controllers.ClusterGroupUpgrade	[getClustersBySelectors]	{"clusterNames": ["cnfdf18"]}
2022-06-03T10:52:56.140-0400	INFO	controllers.ClusterGroupUpgrade	[Reconcile]	{"RequeueAfter:": "5m0s"}
2022-06-03T10:52:56.489-0400	ERROR	controller-runtime.manager.controller.clustergroupupgrade	Reconciler error	{"reconciler group": "ran.openshift.io", "reconciler kind": "ClusterGroupUpgrade", "name": "cnfdf18-new", "namespace": "ztp-install", "error": "Operation cannot be fulfilled on clustergroupupgrades.ran.openshift.io \"cnfdf18-new\": the object has been modified; please apply your changes to the latest version and try again"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/jun/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.3-0.20210709165254-650ea59f19cc/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/home/jun/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.3-0.20210709165254-650ea59f19cc/pkg/internal/controller/controller.go:214
```

Reply:
No problem. See the resource version at line 8 and line 40 in this log (generated without the delay).
[log.txt](https://github.com/openshift-kni/cluster-group-upgrades-operator/files/8833426/log.txt)

Reply:
Actually, I just think of this case still can be handled at the first place. To check if resource has already created, we could query with annotation` utils.DesiredResourceName` as original resource name is known.

Reply:
Hmm, I don't think you can query by annotation. Even if we use labels instead, it would still be a slow list op that this PR is meant to avoid. It would be one list op per object vs one per kind (i.e. 3 total).